pyslim/slim_metadata.py

```diff
@@ -11,6 +13,431 @@
+                        "required": ["source_subpop", "migration_rate"],
+                        "additionalProperties": False
+                    }
+                }
+            },
+            "required": ["slim_id", "selfing_fraction", "female_cloning_fraction",
+                         "male_cloning_fraction", "sex_ratio", "bounds_x0", "bounds_x1",
+                         "bounds_y0", "bounds_y1", "bounds_z0", "bounds_z1", "migration_records"],
+            "additionalProperties": False,
+            "examples": [
+                {
+                   "slim_id": 2,
+                   "selfing_fraction": 0.5,
+                   "female_cloning_fraction": 0.25,
+                   "male_cloning_fraction": 0.0,
+                   "sex_ratio": 0.5,
+                   "bounds_x0": 0.0,
+                   "bounds_x1": 100.0,
+                   "bounds_y0": 0.0,
+                   "bounds_y1": 100.0,
+                   "bounds_z0": 0.0,
+                   "bounds_z1": 100.0,
+                   "migration_records": [
+                       {"source_subpop": 1,
+                        "migration_rate": 0.9 },
+                       {"source_subpop": 2,
+                        "migration_rate": 0.1 }
+                   ]
+                }
+            ]
+        },
+    }
+
+
+slim_metadata_schemas = {k: tskit.MetadataSchema(_raw_slim_metadata_schemas[k])
+        for k in _raw_slim_metadata_schemas}
+
+
+default_slim_metadata = {
+    "tree_sequence" : {
+         "SLiM" : {
+             "model_type" : "nonWF",
+             "generation" : 0,
+             "file_version" : slim_file_version,
+             "spatial_dimensionality" : "",
+             "spatial_periodicity" : "",
+             "separate_sexes" : False,
+             "nucleotide_based" : False,
+         }
+    },
+    "edge" : None,
+    "site" : None,
+    "mutation" : {
+        "mutation_list": []
+    },
+    "node" : {
+        "slim_id": tskit.NULL,
+        "is_null": 1,
+        "genome_type": 0,
+    },
+    "individual" : {
+        "pedigree_id": tskit.NULL,
+        "age": -1,
+        "subpopulation": tskit.NULL,
+        "sex": -1,
+        "flags": 0,
+    },
+    "population" : {
+        "slim_id": tskit.NULL,
+        "selfing_fraction": 0.0,
+        "female_cloning_fraction": 0.0,
+        "male_cloning_fraction": 0.0,
+        "sex_ratio": 0.5,
+        "bounds_x0": 0.0,
+        "bounds_x1": 0.0,
+        "bounds_y0": 0.0,
+        "bounds_y1": 0.0,
+        "bounds_z0": 0.0,
+        "bounds_z1": 0.0,
+        "migration_records": []
+    },
+}
+
+###########
+# Top-level, a.k.a., tree sequence metadata
+###########
+
+def set_tree_sequence_metadata(tables,
+        model_type,
+        slim_generation,
+        spatial_dimensionality='',
+        spatial_periodicity='',
+        separate_sexes=False,
+        nucleotide_based=False):
+    if tables.metadata_schema.schema is None:
+        if len(tables.metadata) > 0:
```

Comment:
Probably best here to use `isinstance(tables.metadata, bytes)` to test for raw metadata.